### PR TITLE
Fixed NMR implementation and examples

### PR DESCRIPTION
**Pull request by TommyX12**
Fixed NMR implementation in Kaolin core (in `kaolin/graphics`). Refactored NMR examples (in `examples/renderers/NMR`) to be more organized and user-friendly, and to make use of Kaolin. Added tests for the examples.

Tested on Ubuntu 16.04, CUDA 10.0, PyTorch 1.3.1.

Other fixes:
- Fixed several tests to use new KaolinDatabase syntax.

### CODE REVIEW

File: examples/renderers/NMR/README.md [hunk 1]
```diff
@@ -1,64 +1,60 @@
...
-python ./examples/example2.py
-python ./examples/example3.py
-python ./examples/example4.py
-```
+To run each example, simply execute `python <example-script>`, where `<example-script>` is one of the four examples in this directory.
 
+To see the command line arguments of each example, run `python <example-script> --help`.
 
-## Example 1: Drawing an object from multiple viewpoints
+(Note: for Python 3, you might need to use `python3` instead.)
```
Comment (Caenorst) [inline examples/renderers/NMR/README.md]: Kaolin is python3 only, so this is unecessary note

File: examples/renderers/NMR/example1.py [hunk 2]
```diff
@@ -0,0 +1,121 @@
+# Copyright (c) 2019, NVIDIA CORPORATION. All rights reserved.
```
Comment (Caenorst) [inline examples/renderers/NMR/example1.py]: copyright year should now be 2020

File: examples/renderers/NMR/example2.py [hunk 3]
```diff
@@ -0,0 +1,204 @@
+# Copyright (c) 2019, NVIDIA CORPORATION. All rights reserved.
```
Comment (Caenorst) [inline examples/renderers/NMR/example2.py]: copyright year should now be 2020

File: examples/renderers/NMR/example1.py [hunk 4]
```diff
@@ -0,0 +1,121 @@
...
+# AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+# LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+# OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
+# SOFTWARE.
+
+from kaolin.graphics import NeuralMeshRenderer as Renderer
+from kaolin.graphics.nmr.util import get_points_from_angles
+from kaolin.rep import TriangleMesh
+from util import normalize_vertices
+import argparse
```
Comment (Caenorst) [inline examples/renderers/NMR/example1.py:41]: It's usually a convention to first import builtin modules (https://www.python.org/dev/peps/pep-0008/#id23)

File: examples/renderers/NMR/example1.py [hunk 5]
```diff
@@ -0,0 +1,121 @@
...
+def main():
+    args = parse_arguments()
+
+    ###########################
+    # Load mesh
+    ###########################
+
+    mesh = TriangleMesh.from_obj(args.mesh)
+    # Normalize into unit cube, and expand such that batch size = 1
+    vertices = normalize_vertices(mesh.vertices).unsqueeze(0).cuda()
```
Comment (Caenorst) [inline examples/renderers/NMR/example1.py]: why is mesh.vertices not converted to cuda before being passed to normalize_vertices ?

File: examples/renderers/NMR/example1.py [hunk 6]
```diff
@@ -0,0 +1,121 @@
...
+    faces = mesh.faces.unsqueeze(0).cuda()
+
+    ###########################
+    # Generate texture (NMR format)
+    ###########################
+
+    textures = torch.ones(
+        1, faces.shape[1], TEXTURE_SIZE, TEXTURE_SIZE, TEXTURE_SIZE,
+        3, dtype=torch.float32
+    ).cuda()
```
Comment (Caenorst) [inline examples/renderers/NMR/example1.py]: use argument `device` in torch.ones instead

File: examples/renderers/NMR/example1.py [hunk 7]
```diff
@@ -0,0 +1,121 @@
...
+    os.makedirs(args.output_path, exist_ok=True)
+    writer = imageio.get_writer(os.path.join(
+        args.output_path, 'example1.gif'), mode='I')
+    for azimuth in loop:
+        renderer.eye = get_points_from_angles(
+            CAMERA_DISTANCE, ELEVATION, azimuth)
+
+        images, _, _ = renderer(vertices, faces, textures)
+
+        image = images.detach().cpu().numpy()[0].transpose(
```
Comment (Caenorst) [inline examples/renderers/NMR/example1.py]: May not be a big deal but you can use `permute` on gpu before transfering to cpu

File: examples/renderers/NMR/example1.py [hunk 8]
```diff
@@ -0,0 +1,121 @@
...
+
+ROOT_DIR = os.path.abspath(os.path.dirname(__file__))
+
+###########################
+# Settings
+###########################
+
+CAMERA_DISTANCE = 2.732
+ELEVATION = 30
+TEXTURE_SIZE = 2
```
Comment (Caenorst) [inline examples/renderers/NMR/example1.py]: Shouldn't those be in args ?

File: examples/renderers/NMR/example2.py [hunk 9]
```diff
@@ -0,0 +1,204 @@
...
+ROOT_DIR = os.path.abspath(os.path.dirname(__file__))
+
+###########################
+# Settings
+###########################
+
+CAMERA_DISTANCE = 2.732
+ELEVATION = 0
+AZIMUTH = 90
+TEXTURE_SIZE = 2
```
Comment (Caenorst) [inline examples/renderers/NMR/example2.py]: Shouldn't those be in args ?

File: examples/renderers/NMR/example2.py [hunk 10]
```diff
@@ -0,0 +1,204 @@
...
+
+    ###########################
+    # Optimize
+    ###########################
+
+    loop = tqdm.tqdm(range(args.epochs))
+    loop.set_description('Optimizing')
+
+    optimizer = torch.optim.Adam(
+        filter(lambda p: p.requires_grad, model.parameters()))
```
Comment (Caenorst) [inline examples/renderers/NMR/example2.py]: List comprehension is preferable

File: examples/renderers/NMR/example3.py [hunk 11]
```diff
@@ -0,0 +1,227 @@
...
+
+ROOT_DIR = os.path.abspath(os.path.dirname(__file__))
+
+###########################
+# Settings
+###########################
+
+CAMERA_DISTANCE = 2.732
+ELEVATION = 0
+TEXTURE_SIZE = 4
```
Comment (Caenorst) [inline examples/renderers/NMR/example3.py]: Shouldn't those be in args ?

File: examples/renderers/NMR/example3.py [hunk 12]
```diff
@@ -0,0 +1,227 @@
+# Copyright (c) 2019, NVIDIA CORPORATION. All rights reserved.
```
Comment (Caenorst) [inline examples/renderers/NMR/example3.py]: copyright year should now be 2020

File: examples/renderers/NMR/example3.py [hunk 13]
```diff
@@ -0,0 +1,227 @@
...
+
+    ###########################
+    # Optimize
+    ###########################
+
+    loop = tqdm.tqdm(range(args.epochs))
+    loop.set_description('Optimizing')
+
+    optimizer = torch.optim.Adam(
+        filter(lambda p: p.requires_grad, model.parameters()),
```
Comment (Caenorst) [inline examples/renderers/NMR/example3.py]: List comprehension is preferable

File: examples/renderers/NMR/example2.py [hunk 14]
```diff
@@ -0,0 +1,204 @@
...
+
+        loss = model()
+
+        loss.backward()
+        optimizer.step()
+
+        images, _, _ = model.renderer(
+            model.vertices, model.faces, model.textures)
+
+        image = images.detach().cpu().numpy()[0].transpose(
```
Comment (Caenorst) [inline examples/renderers/NMR/example2.py]: May not be a big deal but you can use `permute` on gpu before transfering to cpu

File: examples/renderers/NMR/example2.py [hunk 15]
```diff
@@ -0,0 +1,204 @@
...
+    writer = imageio.get_writer(os.path.join(
+        args.output_path, 'example2_mesh.gif'), mode='I')
+    for azimuth in loop:
+        model.renderer.eye = get_points_from_angles(
+            CAMERA_DISTANCE, ELEVATION, azimuth)
+
+        images, _, _ = model.renderer(
+            model.vertices, model.faces, model.textures)
+
+        image = images.detach().cpu().numpy()[0].transpose(
```
Comment (Caenorst) [inline examples/renderers/NMR/example2.py]: May not be a big deal but you can use `permute` on gpu before transfering to cpu

File: examples/renderers/NMR/example3.py [hunk 16]
```diff
@@ -0,0 +1,227 @@
...
+        model.renderer.eye = get_points_from_angles(
+            CAMERA_DISTANCE, ELEVATION, azimuth)
+
+        images, _, _ = model.renderer(
+            model.vertices,
+            model.faces,
+            torch.tanh(model.textures)
+        )
+
+        image = images.detach().cpu().numpy()[0].transpose(
```
Comment (Caenorst) [inline examples/renderers/NMR/example3.py]: May not be a big deal but you can use `permute` on gpu before transfering to cpu

File: examples/renderers/NMR/example3.py [hunk 17]
```diff
@@ -0,0 +1,227 @@
...
```
Comment (Caenorst) [inline examples/renderers/NMR/example3.py]: May not be a big deal but you can use `permute` on gpu before transfering to cpu

File: examples/renderers/NMR/example4.py [hunk 18]
```diff
@@ -0,0 +1,196 @@
+# Copyright (c) 2019, NVIDIA CORPORATION. All rights reserved.
```
Comment (Caenorst) [inline examples/renderers/NMR/example4.py]: copyright year should now be 2020

File: examples/renderers/NMR/example4.py [hunk 19]
```diff
@@ -0,0 +1,196 @@
...
+ROOT_DIR = os.path.abspath(os.path.dirname(__file__))
+
+###########################
+# Settings
+###########################
+
+INITIAL_CAMERA_POS = [6, 10, -14]
+CAMERA_DISTANCE = 2.732
+ELEVATION = 0
+TEXTURE_SIZE = 2
```
Comment (Caenorst) [inline examples/renderers/NMR/example4.py]: Shouldn't those be in args ?

File: examples/renderers/NMR/example4.py [hunk 20]
```diff
@@ -0,0 +1,196 @@
...
+
+    ###########################
+    # Optimize
+    ###########################
+
+    loop = tqdm.tqdm(range(args.epochs))
+    loop.set_description('Optimizing')
+
+    optimizer = torch.optim.Adam(
+        filter(lambda p: p.requires_grad, model.parameters()),
```
Comment (Caenorst) [inline examples/renderers/NMR/example4.py]: List comprehension is preferable

File: examples/renderers/NMR/example4.py [hunk 21]
```diff
@@ -0,0 +1,196 @@
...
+        loss.backward()
+        optimizer.step()
+
+        images, _, _ = model.renderer(
+            model.vertices,
+            model.faces,
+            model.textures
+        )
+
+        image = images.detach().cpu().numpy()[0].transpose(
```
Comment (Caenorst) [inline examples/renderers/NMR/example4.py]: May not be a big deal but you can use `permute` on gpu before transfering to cpu

**Review by Caenorst — CHANGES_REQUESTED** [review]
Looking good overall, only changing copyright year and replacing filter by list comprehension is required, rest is recommended only

File: kaolin/graphics/NeuralMeshRenderer.py [hunk 22]
```diff
@@ -36,529 +36,265 @@
...
-            viewing_angle: float = 30.,
-            camera_direction: torch.Tensor = torch.FloatTensor([0, 0, 1]),
-            near: float = 0.1, far: float = 100,
-            light_intensity_ambient: float = 0.5,
-            light_intensity_directional: float = 0.5,
-            light_color_ambient: torch.Tensor = torch.ones(3),
-            light_color_directional: torch.Tensor = torch.ones(3),
-            light_direction: torch.Tensor = torch.FloatTensor([0, 1, 0]),
-            device: str = 'cpu'):
-        r"""Initalize the NeuralMeshRenderer object.
```
Comment (Caenorst) [inline kaolin/graphics/NeuralMeshRenderer.py:76]: Why removing all the doc ?

File: kaolin/graphics/NeuralMeshRenderer.py [hunk 23]
```diff
@@ -36,529 +36,265 @@
...
+
+class NeuralMeshRenderer(nn.Module):
+    def __init__(self, image_size=256, anti_aliasing=True, background_color=[0, 0, 0],
+                 fill_back=True, camera_mode='projection',
+                 K=None, R=None, t=None, dist_coeffs=None, orig_size=1024,
+                 perspective=True, viewing_angle=30, camera_direction=[0, 0, 1],
+                 near=0.1, far=100,
+                 light_intensity_ambient=0.5, light_intensity_directional=0.5,
+                 light_color_ambient=[1, 1, 1], light_color_directional=[1, 1, 1],
+                 light_direction=[0, 1, 0]):
```
Comment (Caenorst) [inline kaolin/graphics/NeuralMeshRenderer.py]: Please don't put any mutable variable (list here) as default argument.

Comment (Caenorst) [reply kaolin/graphics/NeuralMeshRenderer.py]: If you want those list as default argument, use `None` and set them inside the function.
(Use the final default argument in the doc)

File: kaolin/graphics/NeuralMeshRenderer.py [hunk 24]
```diff
@@ -36,529 +36,265 @@
...
-        # extrinsics.
         if self.camera_mode == 'projection':
             self.K = K
-            self.rmat = rmat
-            self.tvec = tvec
-        # If the mode is 'look' or 'look_at', use the viewing angle to determine
-        # perspective distortion and camera position and orientation.
+            self.R = R
+            self.t = t
+            if isinstance(self.K, numpy.ndarray):
```
Comment (Caenorst) [inline kaolin/graphics/NeuralMeshRenderer.py:158]: Hum, what if it's a tensor but on cpu ? (Should we support this case ?)

File: kaolin/graphics/NeuralMeshRenderer.py [hunk 25]
```diff
@@ -36,529 +36,265 @@
...
+                self.K = torch.cuda.FloatTensor(self.K)
+            if isinstance(self.R, numpy.ndarray):
+                self.R = torch.cuda.FloatTensor(self.R)
+            if isinstance(self.t, numpy.ndarray):
+                self.t = torch.cuda.FloatTensor(self.t)
+            self.dist_coeffs = dist_coeffs
+            if dist_coeffs is None:
+                self.dist_coeffs = torch.cuda.FloatTensor(
```
Comment (Caenorst) [inline kaolin/graphics/NeuralMeshRenderer.py]: torch.zeros

File: kaolin/graphics/NeuralMeshRenderer.py [hunk 26]
```diff
@@ -36,529 +36,265 @@
...
         else:
-            raise ValueError('Mode {0} not implemented.'.format(mode))
-
-    def render(self, vertices, faces, textures=None, mode=None, K=None,
-               rmat=None, tvec=None):
-        r"""Renders the RGB, depth, and alpha channels.
+            raise ValueError(
+                "mode should be one of None, 'silhouettes' or 'depth'")
 
-        Args:
```
Comment (Caenorst) [inline kaolin/graphics/NeuralMeshRenderer.py:223]: You removed the doc but didn't put it back.

**Review by Caenorst — CHANGES_REQUESTED** [review]

**Review by Caenorst — APPROVED** [review]
LGTM, thanks for the contribution ! :)
